cpp/include/raft/core/resource/custom_resource.hpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/resource/resource_types.hpp>
+#include <raft/core/resources.hpp>
+
+#include <memory>
+#include <typeindex>
+
+namespace raft::resource {
+
+class custom_resource : public resource {
+ public:
+  custom_resource()                    = default;
+  ~custom_resource() noexcept override = default;
+  auto get_resource() -> void* override { return this; }
+
+  template <typename ResourceT>
+  auto load() -> ResourceT*
+  {
+    std::lock_guard<std::mutex> _(lock_);
+    auto key = std::type_index{typeid(ResourceT)};
+    auto pos = map_.find(key);
+    if (pos != map_.end()) { return reinterpret_cast<ResourceT*>(pos->second.get()); }
+    auto store_ptr = new ResourceT{};
+    map_[key] =
+      std::shared_ptr<void>(store_ptr, [](void* ptr) { delete reinterpret_cast<ResourceT*>(ptr); });
+    return store_ptr;
+  }
+
+ private:
+  std::unordered_map<std::type_index, std::shared_ptr<void>> map_{};
```

Comment:
@wphicks, as per our discussion today, I was going to replace this with a vector of tuples, but now I'm not sure. We've discussed a little earlier with @cjnolet that the `get_resource` is not likely to be a bottleneck in any of our use cases. Wouldn't it be a little premature optimization to implement this now? Iterating through a vector to find an element and all that - could be a little less readable vs map_.find

Reply:
Update: changed this to the vector as suggested. After some testing, it turns out it does make a bit of a difference!